src/contract_client.rs

```diff
@@ -668,6 +677,193 @@ impl ModuleDeployHandle {
+///
+/// Disclaimer: A smart contract can have logic to overwrite/change the meaning
+/// of the error codes as defined in the `concordium-std` crate. While it is not
+/// advised to overwrite these error codes and is rather unusual to do so, this
+/// function decodes the error codes based on the
+/// definitions in the `concordium-std` crate (assuming they have not been
+/// overwritten with other meanings in the smart contract logic). No guarantee
+/// are given as such that the meaning of the decoded reject reason haven't been
+/// altered by the smart contract logic.
+pub fn decode_smart_contract_revert(
+    return_value: Option<ReturnValue>,
+    reject_reason: RejectReason,
+    schema: &Option<VersionedModuleSchema>,
+    receive_name: OwnedReceiveName,
+    contract_name: &OwnedContractName,
+) -> Result<Option<String>, RejectReasonDecodeError> {
+    match reject_reason {
+        RejectReason::RejectedReceive {
+            reject_reason: reject_reason_code,
+            contract_address: _,
+            receive_name: _,
+            parameter: _,
+        } => {
+            // Step 1: Try to decode the `reject_reason` using the `concordium-std`
+            // error codes.
+            if let Some(decoded_error) = decode_concordium_std_error(reject_reason_code) {
+                return Ok(Some(decoded_error));
+            }
+
+            // Step 2: Try to decode the `reject_reason` using the `error_schema` and the
+            // `return_value`.
+            if let Some(schema) = schema {
+                if let (Some(error_schema), Some(return_value)) = (
+                    get_error_schema(schema, receive_name, contract_name),
+                    return_value,
+                ) {
+                    let mut cursor = Cursor::new(return_value.value);
+
+                    match error_schema.to_json(&mut cursor) {
+                        Ok(serde_json::Value::Object(obj)) => {
+                            if let Some(key) = obj.keys().next() {
+                                return Ok(Some(key.to_string()));
+                            }
+                            // This error can not happen if a valid `error_schema` is provided.
+                            return Err(RejectReasonDecodeError::MissingErrorVariant);
+                        }
+                        Ok(_) => {
+                            // This error can not happen if a valid `error_schema` is provided.
+                            return Err(RejectReasonDecodeError::NoMatchingErrorVariant);
+                        }
+                        // This error can not happen if a valid `error_schema` is provided.
+                        Err(e) => {
+                            return Err(e.into());
+                        }
+                    }
+                }
+            }
+
+            // If no `error_schema` and/or `return_value` is provided, the `reject_reason`
+            // can not be decoded.
+            Ok(None)
```

Comment:
Would flip this on its head to decrease nesting (and IMO make it easier to read):

```rust
let Some(schema) = schema else { return Ok(None) };
let (Some(error_schema), Some(return_value)) = (
    get_error_schema(schema, receive_name, contract_name), return_value,
) else { return Ok(None) };

...
```

Reply:
The order has changed now:
- The smart contract errors take precedence over the `concordium-std` error.

As such it is not possible to return anymore early if no schema is present (since we still need to check for the `concordium-std` error).